docs/GDPs/gdp-0002.rst

```diff
@@ -0,0 +1,342 @@
+        if grid.north_edge:
+            y_edge_ub(ut, ub, dt4=dt4, origin=(grid.local_istart, grid.local_jend, 0), domain=domain_y_edge)
+        if grid.east_edge:
+            x_edge_ub(ut, ub, dt5=dt5, origin=(grid.local_iend, grid.local_jstart, 0), domain=domain_x_edge)
+
+This example motivates a number of requirements for the feature. It must be able to
+
+1. specify regions in the context of a global iteration space (for global boundary conditions) using different `with` blocks inside a single stencil
+2. specify arbitrary sub-regions using a clean abstraction
+3. specify cleanly regions at and near the edge of compute domains
+4. read from and write to fields inside halo regions
+
+In the `detailed description` section below we offer a few alternatives, but all these involve adding a `region()` specifier that allows for specifying subsets of the horizontal iteration space.
+Using this, the example above is transformed into:
+
+.. code-block:: python
+
+    @gtscript.stencil(backend=backend_type, partitioner=part_type) # Single stencil
+    def ubke(uc: FIELD_TYPE, vc: FIELD_TYPE, cosa: FIELD_TYPE, rsina: FIELD_TYPE, ut: FIELD_TYPE, ub: FIELD_TYPE, dt5: float, dt4: float):
+        with computation(PARALLEL):
+            with interval(...): # applies to the compute domain as it does now
+                ub = dt5 * (uc[0, -1, 0] + uc - (vc[-1, 0, 0] + vc) * cosa) * rsina
+            with region((WestEdge(0, 1), EastEdge(0, 1)), interval(...):
+                # Executes on regions defined by the Box objects
+                ub = dt5 * (ut[0, -1, 0] + ut)​
+            with region((SouthEdge, 0, 1), (NorthEdge, 0, 1)), interval(...):
+                ub = dt4 * (-ut[0, -2, 0] + 3.0 * (ut[0, -1, 0] + ut) - ut[0, 1, 0])
+
+    # Python wrapper to unpack arguments
+    def compute(uc, vc, ut, ub, dt5, dt4, grid):​
+        ubke(uc, vc, grid.cosa, grid.rsina, ub, ut, dt5, dt4)
+
+This drastically reduces the complexity of the code and consolidates operations on `ub` - it is now clear what the stencil is doing to `ub` everywhere.
+Model developers already understand the idea of offsets with the stencil notation of e.g. `uc[0, -1, 0]`, which is analogous to specifying offsets from an edge of the horizontal domain.
+
+All that needs to be changed in the frontend is adding the `region()` specifier, **so all existing GT4Py code will continue to work as before.**
+The new `region()` specifier accepts what are essentially slices that describe the horizontal iteration space in a way that describes the process' local iteration space, and can also represent the global space if the space is domain decomposed and computed by a set of processes.
+
+It may not be immediately obvious that this supports the notion of a global iteration space, but the object passed to `region()` is constructed by the specific application, and can embed details about the domain partitioning.
```

Comment:
Most of the time gt4py seems to say "user" when they talk about the application, I think.
```suggestion
It may not be immediately obvious that this supports the notion of a global iteration space, but the object passed to `region()` is constructed by the user, and can embed details about the domain partitioning. It is the responsibility of this object to translate from a global index provided by the user to a local index used by gt4py.
```